File: 2021/23.py
dat = ["DACA","DCBB"]

POS = [0,1,3,5,7,9,10]

class State:
	def __init__(self, depth, map):
		self.depth = depth
		self.map = tuple(tuple(i) if not isinstance(i,tuple) else i for i in map)
	def __repr__(self):
		return f"State({self.depth}, {self.map})"
	def __str__(self):
		return f"#############\n#{self._chr(4,0)}{self._chr(4,1)}.{self._chr(4,2)}.{self._chr(4,3)}.{self._chr(4,4)}.{self._chr(4,5)}{self._chr(4,6)}#\n" + "".join(
			f"{'##' if i == 0 else '  '}#{self._chr(0,i)}#{self._chr(1,i)}#{self._chr(2,i)}#{self._chr(3,i)}#{'##' if i == 0 else '  '}\n" for i in range(self.depth)
		) + "  #########  "
	def _chr(self, a, b):
		if self.map[a][b] is None:
			return "."
		else:
			return chr(self.map[a][b] + 65)
	def __hash__(self):
		return hash(self.map)
	def __eq__(self, other):
		return self.map == other.map
	def neighbours(self):
		for i in range(4):
			for j in range(self.depth):
				if self.map[i][j] is not None:
					break
			else:
				continue
			for k in range(i+1,-1,-1):
				if self.map[4][k] is not None:
					break
				yield self.newstate(i,j,4,k)
			for k in range(i+2,7):
				if self.map[4][k] is not None:
					break
				yield self.newstate(i,j,4,k)
		for i in range(7):
			if self.map[4][i] is None:
				continue
			target = self.map[4][i]
			if not all(j is None or j == target for j in self.map[target]):
				continue
			if i <= target+1:
				if not all(j is None for j in self.map[4][i+1:target+2]):
					continue
			else:
				if not all(j is None for j in self.map[4][target+2:i]):
					continue
			k = -1
			while k+1 < self.depth and self.map[target][k+1] is None:
				k += 1
			if k < 0:
				continue
			yield self.newstate(4,i,target,k)
	@staticmethod
	def dist(froma,fromb,toa,tob):
		if froma == 4 and toa == 4:
			return abs(POS[fromb] - POS[tob])
		elif froma == 4:
			return abs(POS[fromb] - 2*(toa+1)) + tob+1
		elif toa == 4:
			return fromb+1 + abs(2*(froma+1) - POS[tob])
		else:
			return fromb+1 + abs(2*(froma+1) - 2*(toa+1)) + tob+1
	def newstate(self,froma,fromb,toa,tob):
		newmap = list(self.map)
		newmap[froma] = list(newmap[froma])
		newmap[toa] = list(newmap[toa])
		weight = newmap[toa][tob] = newmap[froma][fromb]
		newmap[froma][fromb] = None
		return State(self.depth, newmap), self.dist(froma,fromb,toa,tob) * 10**weight
	def isfinal(self):
		return all(self.map[i][j] == i for i in range(4) for j in range(self.depth)) and all(self.map[4][i] is None for i in range(7))
	def heuristic(self):
		def gen():
			seen = [0,0,0,0]
			for i in range(4):
				j = self.depth - 1
				while j >= 0 and self.map[i][j] == i:
					seen[i] += 1
					j -= 1
			for i,group in enumerate(self.map):
				for j,x in enumerate(group):
					if x is not None and (i == 4 or not all(self.map[i][k] == i for k in range(j,self.depth))):
						seen[x] += 1
						yield self.dist(i,j,x,self.depth-seen[x]) * 10**x
		return sum(gen())

def make_initstate(dat):
	dat = [[ord(i) - 65 for i in j] for j in dat]
	return State(len(dat), (*zip(*dat), [None] * 7))

def go_dijkstra(initstate):
	statemap = {initstate: 0}
	candidates = {initstate}
	while candidates:
		minstate = min(candidates, key=statemap.__getitem__)
		if minstate.isfinal():
			return statemap[minstate]
		candidates.remove(minstate)
		nodecost = statemap[minstate]
		for n, c in minstate.neighbours():
			if n not in statemap or statemap[n] > nodecost + c:
				statemap[n] = nodecost + c
				candidates.add(n)
	else:
		return None

def go_astar(initstate):
	statemap = {initstate: 0}
	score = {initstate: initstate.heuristic()}
	candidates = {initstate}
	while candidates:
		minstate = min(candidates, key=score.__getitem__)
		if minstate.isfinal():
			return statemap[minstate]
		candidates.remove(minstate)
		nodecost = statemap[minstate]
		for n, c in minstate.neighbours():
			if n not in statemap or statemap[n] > nodecost + c:
				statemap[n] = nodecost + c
				score[n] = nodecost + c + n.heuristic()
				candidates.add(n)
	else:
		return None

#print(go_dijkstra(make_initstate(dat)))
print(go_astar(make_initstate(dat)))
print(go_astar(make_initstate([dat[0], "DCBA", "DBAC", dat[1]])))

File: 2021/23.hs
{-# OPTIONS_GHC -Wno-tabs #-}
import Data.Char
import Data.List
import Data.Maybe
import qualified Data.Map.Strict as M
import Control.Exception
import Control.Monad
import Dijkstra

type State = [[Maybe Int]]

makeInitState :: [String] -> State
makeInitState rows = (transpose $ map (map fromLetter) rows) ++ [[Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing]]
	where fromLetter c = Just $ ord c - ord 'A'

showState :: State -> String
showState state = unlines $ [toprow, hallway] ++ [home i | i <- [0..length (head state) - 1]] ++ [bottomrow]
	where
		toprow = "#############"
		hallway = ['#', c 4 0, c 4 1, ' ', c 4 2, ' ', c 4 3, ' ', c 4 4, ' ', c 4 5, c 4 6, '#']
		home n = [b, b, '#', c 0 n, '#', c 1 n, '#', c 2 n, '#', c 3 n, '#', b, b]
			where b = if n == 0 then '#' else ' '
		bottomrow = "  #########  "
		c n m = case state !! n !! m of Nothing -> ' '; (Just x) -> chr (x + ord 'A')

hallwaypos :: Int -> Int
hallwaypos 0 = 0
hallwaypos 1 = 1
hallwaypos 2 = 3
hallwaypos 3 = 5
hallwaypos 4 = 7
hallwaypos 5 = 9
hallwaypos 6 = 10

calcdist :: Int -> Int -> Int -> Int -> Integer
calcdist 4 fromb 4 tob = toInteger $ abs (hallwaypos fromb - hallwaypos tob)
calcdist froma fromb 4 tob = toInteger $ fromb+1 + abs (2*(froma+1) - hallwaypos tob)
calcdist 4 fromb toa tob = toInteger $ abs (hallwaypos fromb - 2*(toa+1)) + tob+1
calcdist froma fromb toa tob = toInteger $ fromb+1 + abs (2*(froma+1) - 2*(toa+1)) + tob+1

setval :: Int -> a -> [a] -> [a]
setval _ _ [] = []
setval 0 x (_:rest) = x:rest
setval n x (a:rest) = a:setval (n-1) x rest

setval2 :: Int -> Int -> a -> [[a]] -> [[a]]
setval2 _ _ _ [] = []
setval2 0 m x (a:rest) = setval m x a:rest
setval2 n m x (a:rest) = a:setval2 (n-1) m x rest

nextState :: Int -> Int -> Int -> Int -> State -> (State, (), Integer)
nextState froma fromb toa tob state = (setval2 froma fromb Nothing $ setval2 toa tob (Just target) state, (), calcdist froma fromb toa tob * 10^target)
	where (Just target) = state !! froma !! fromb

neighbours :: State -> [(State, (), Integer)]
neighbours state = join $ [moveOut state i | i <- [0..3]] ++ [moveHome state i | i <- [0..6]]

moveOut :: State -> Int -> [(State, (), Integer)]
moveOut state froma = if homeReady then [] else map makeState $ targetleft ++ targetright
	where
		home = state !! froma
		hallway = state !! 4
		homeReady = all (\x -> isNothing x || x == (Just froma)) home
		fromb = length $ takeWhile isNothing home
		targetleft = takeWhile (isNothing . (hallway !!)) [froma+1, froma .. 0]
		targetright = takeWhile (isNothing . (hallway !!)) [froma+2 .. 6]
		makeState tob = nextState froma fromb 4 tob state

moveHome :: State -> Int -> [(State, (), Integer)]
moveHome state fromb = if cellEmpty then [] else if not hallwayReady || not homeReady then [] else [nextState 4 fromb toa tob state]
	where
		hallway = state !! 4
		cellEmpty = isNothing $ hallway !! fromb
		(Just toa) = hallway !! fromb
		home = state !! toa
		hallwayThrough = if toa+1 > fromb then [fromb+1..toa+1] else [toa+2..fromb-1]
		hallwayReady = all isNothing $ map (hallway !!) hallwayThrough
		homeReady = all (\x -> isNothing x || x == (Just toa)) home
		tob = subtract 1 $ length $ takeWhile isNothing home

heuristic :: State -> Integer
heuristic state = fromHallway + fromWrongHome + furtherDepth
	where
		fromHallway = sum $ map fromHallwayCell [0..6]
		fromHallwayCell n = case state !! 4 !! n of Nothing -> 0; (Just x) -> calcdist 4 n x 0 * 10^x
		depth = length $ head state
		notHome n = depth - (length $ takeWhile (==(Just n)) $ reverse (state !! n))
		notHomeCount = map notHome [0..3]
		fromWrongHome = sum [ fromWrongHomeCell a b | a <- [0..3], b <- [0..(notHomeCount !! a) - 1]]
		fromWrongHomeCell a b = case state !! a !! b of Nothing -> 0; (Just x) -> calcdist a b x 0 * 10^x
		furtherDepthCol n = let d = notHomeCount !! n in toInteger $ d * (d - 1) `div` 2 * 10^n
		furtherDepth = sum $ map furtherDepthCol [0..3]

isFinalState :: State -> Bool
isFinalState state = all (\(r,n) -> all (==(Just n)) r) $ zip state [0..3]

solveState :: State -> Integer
solveState state = let (n, _, _) = distmap M.! finalstate in n
	where (distmap, Just finalstate) = buildDistMapAstar state neighbours isFinalState heuristic

solvePartA a b = solveState $ makeInitState [a, b]
solvePartB a b = solveState $ makeInitState [a, "DCBA", "DBAC", b]

tests :: IO ()
tests = do
	check $ (solvePartA "BCBD" "ADCA") == 12521
	check $ (solvePartB "BCBD" "ADCA") == 44169
	where
		check True = return ()
		check False = throwIO $ AssertionFailed "test failed"

main :: IO ()
main = do
	--tests
	print $ solvePartA "DACA" "DCBB"
	print $ solvePartB "DACA" "DCBB"
